### PR TITLE
Implement the Mahalanobis distance (#6)

### PR DESCRIPTION
**Pull request by thevolatilebit**
Added the option to use a Mahalanobis distance to compute distance between readout vectors. Update tests.

Also updated `DistanceBased` so that it is no more assumed that the distance is just the sum of column-wise distances. Keyword `distances` replaced by `distance`.

Prioritizing keyword arguments for method signatures in `GenerativeDesigns`.

### CODE REVIEW

File: src/GenerativeDesigns/distancebased.jl [hunk 1]
```diff
@@ -50,20 +52,97 @@ end
...
+# Arguments
+
+  - `diagonal`: A scalar or vector to be added to the diagonal entries of the covariance matrix.
+
+# Returns
+
+It returns a high-level function of `(data, targets, prior)`.
+When called, that function will return an internal function `compute_distances` that takes an `Evidence` and computes the Mahalanobis distance based on the input data and the evidence.
+"""
+function MahalanobisDistance(; diagonal = 0)
```
Comment (danielchen26) [inline src/GenerativeDesigns/distancebased.jl:101]:
```julia
using Distances
using LinearAlgebra
using StatsBase

function mahalanobisdistance(; diagonal=0)
    function (data, targets, prior)
        non_targets = setdiff(names(data), targets)
        # Use DataFrame accessors for columns and convert them to Matrix if needed.
        data_matrix = convert(Matrix, data[:, non_targets])
        # Calculate the weighted covariance matrix and adjust the diagonal if specified.
        σ = cov(data_matrix, Weights(prior))
        if diagonal != 0
            if diagonal isa Number
                σ += I * diagonal
            else
                @assert length(diagonal) == size(σ, 1) "Length of diagonal must match the dimensions of the covariance matrix."
                diag_indices = diagind(σ)
                σ[diag_indices] += diagonal
            end
        end
        
        # Check if covariance matrix is singular.
        if rank(σ) < length(non_targets)
            error("Singular covariance matrix σ (rank $(rank(σ))). Consider adding entries to diagonal via `diagonal` keyword.")
        end

        # Pre-compute the inverse of the covariance matrix for performance.
        inv_σ = inv(σ)
        
        compute_distances = function (evidence)
            if isempty(evidence)
                return zeros(size(data, 1))
            else
                # Construct a vector for the evidence, with zeros for missing non-target columns.
                vec_evidence = [get(evidence, col, 0) for col in non_targets]
                # Calculate the Mahalanobis distance for each row in the data.
                distances = [mahalanobis(vec_row, vec_evidence, inv_σ) for vec_row in eachrow(data[:, non_targets])]
                return distances
            end
        end
        return compute_distances
    end
end
```
`Distance.jl` has the internal function for the calculation. I refactor it for the row_wise Mahalanobis distance calculation

File: docs/src/tutorials/GenerativeDesigns.jl [hunk 2]
```diff
@@ -249,7 +257,7 @@ designs = efficient_designs(
     6,
     evidence;
     solver,
-    mdp_options = (; max_parallel = 2, costs_tradeoff = [0, 1.0]),
+    mdp_options = (; max_parallel = 2, costs_tradeoff = (0, 1.0)),
```
Comment (danielchen26) [inline docs/src/tutorials/GenerativeDesigns.jl:260]: costs_tradeoff has to be tuple? I somehow remember I just use it as a list before

**Review by danielchen26 — COMMENTED** [review]
@thevolatilebit I have some comments on the function naming and am suggesting a refactored code for Mahalanobis distance. Let me know if it makes sense.

File: docs/src/tutorials/GenerativeDesigns.jl (hunk 2 above)
Comment (thevolatilebit) [reply docs/src/tutorials/GenerativeDesigns.jl:260]: Yeah, we would like to stick to tuples after https://github.com/Merck/CEED.jl/pull/3.

File: src/GenerativeDesigns/distancebased.jl (hunk 1 above)
Comment (thevolatilebit) [reply src/GenerativeDesigns/distancebased.jl:101]: Hmm, I am not sure if it worth to introduce additional dependency, especially considering that the calculation involved by Distances.jl is just `dot(z, Q * z)`.

File: tutorials/GenerativeDesigns.jl [hunk 3]
```diff
@@ -127,7 +129,13 @@ using CEED, CEED.GenerativeDesigns
 (; sampler, uncertainty, weights) =
```
[the remark this answers is not in the document]
Comment (thevolatilebit) [reply tutorials/GenerativeDesigns.jl:129]: Thanks, I changes the signatures of several methods in `GenerativeDesigns` accordingly.

File: src/GenerativeDesigns/distancebased.jl (hunk 1 above)
Comment (thevolatilebit) [reply src/GenerativeDesigns/distancebased.jl:101]: Yeah, it is just a matrix multiplication and a dot product. Updated to use the same code as Distances.jl does.

**Review by danielchen26 — APPROVED** [review]

**Review by danielchen26 — APPROVED** [review]

**Pull request merged**